frame/executive/src/lib.rs

```diff
@@ -176,13 +178,21 @@ where
 		extrinsics_root: &System::Hash,
 		digest: &Digest<System::Hash>,
 	) {
+		<AllModules as OnRuntimeUpgrade>::on_runtime_upgrade();
```

Comment:
Hmm actually in my branch I moved this to under system::initialize, so order would be:

system::initialize
all_modules::on_runtime_upgrade
all_modules::on_initialize

Reply:
`system::initialize` does a bunch of stuff with the storage of the System module:

```
		// populate environment
		storage::unhashed::put(well_known_keys::EXTRINSIC_INDEX, &0u32);
		<Number<T>>::put(number);
		<Digest<T>>::put(digest);
		<ParentHash<T>>::put(parent_hash);
		<BlockHash<T>>::insert(*number - One::one(), parent_hash);
		<ExtrinsicsRoot<T>>::put(txs_root);

		if let InitKind::Full = kind {
			<Events<T>>::kill();
			EventCount::kill();
			<EventTopics<T>>::remove_all();
		}
```

From a principal standpoint, it seems we should do all the `on_runtime_upgrade()` stuff before any module (even the system) does storage things. For example, if we want to use `on_runtime_upgrade()` on the system module itself, right?

Reply:
yes ok, I was thinking `system::initialize` is a bit different from `system::on_initialize` and might be needed for a module to behave correctly even in on_runtime_upgrade. And also in case we wanted system to upgrade we could directly call this upgrade from executive as it is frame-system.

But indeed I think I'm ok with having on_runtime_upgrade prior to system::initialize.

Reply:
There actually is no `system:on_initialize`. We hardcode this `system::initialize. I don't know why, but my guess is so that we can guarantee that our stuff runs first...?

Reply:
I don't know how much it is allowed but construct_runtime seems to accept having system not as first module indeed.

Reply:
IMO should force System to be first :)

Reply:
or we should remove it from the all modules list, and have it be hand-coded into the macros